File: src/start/LeetCode_24_SwapNodeInPairs.java
package start;

import start.struct.ListNode;

class LeetCode_24_SwapNodeInPairs_Solution {

    public ListNode swapPairs(ListNode head) {
        ListNode first = new ListNode(0);
        first.next = head;
        swap(first, head);
        return first.next;
    }

    private void swap(ListNode pre, ListNode cur) {
        if (cur == null || cur.next == null) {
            return ;
        }

        ListNode nx = cur.next;
        pre.next = nx;
        cur.next = nx.next;
        nx.next = cur;

        swap(cur, cur.next);
    }
}

public class LeetCode_24_SwapNodeInPairs {
}

File: src/golang-leetcode/LeetCode_24_SwapNodeInPairs.go
package main

// swapPairs 两两交换连表中的节点
func swapPairs(head *ListNode) *ListNode {
	first := &ListNode{
		Val:  0,
		Next: head,
	}
	swapPairsLogic(first, head)
	return first.Next
}

// swapPairsLogic 主逻辑
func swapPairsLogic(pre *ListNode, cur *ListNode) {
	if cur == nil || cur.Next == nil {
		return
	}

	nx := cur.Next
	pre.Next = nx
	cur.Next = nx.Next
	nx.Next = cur

	swapPairsLogic(cur, cur.Next)
}
